### PR TITLE
support num_sanity_val_steps=-1

### LINKED ISSUE #1715: Run full validation epoch before training
#### What is your question?
How can I manually trigger a full validation step before training?

I want to compute and log the validation metrics before I start the training (ideally also updating the progress bar dictionary).

The reason why I want to do this is that I am fine-tuning a pre-trained model, and I want to check the performances before training.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #1715 
`Trainer(num_sanity_val_steps=-1)` is now possible and will run all val dataloaders in full.

TODO: 
- [x] is `limit_val_batches` supposed to influence sanity val steps? ok, clarified by william

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@williamFalcon is `limit_val_batches=0` supposed to influence `num_sanity_val_steps`, i.e., should 
`Trainer(num_sanity_val_steps=5, limit_val_batches=0)` run 5 checks or none? Currently it runs none.

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -804,6 +805,24 @@ def test_tpu_choice(tmpdir, tpu_cores, expected_tpu_id, error_expected):
...
 
+def test_num_sanity_val_steps(tmpdir):
+    """ Test that num_sanity_val_steps=-1 runs through all validation data once. """
+    model = EvalModelTemplate()
+    trainer = Trainer(
+        num_sanity_val_steps=-1,
+        # TODO: limit_val_batches influences num_sanity_val_step. Fix it.
+        # limit_val_batches=0,
+        max_steps=1,
+        default_root_dir=tmpdir
```
Comment (Borda) [inline tests/trainer/test_trainer.py:816]:
```suggestion
        default_root_dir=tmpdir,
```

**Comment by Borda** [comment]
> is limit_val_batches supposed to influence sanity val steps?

it depends if you shuffle them... @williamFalcon?

**Comment by williamFalcon** [comment]
> @williamFalcon is `limit_val_batches=0` supposed to influence `num_sanity_val_steps`, i.e., should
> `Trainer(num_sanity_val_steps=5, limit_val_batches=0)` run 5 checks or none? Currently it runs none.

yup it should run none.
Those are two independent actions.

sanity check happens at every run and it is a constant (tiny) number.
The other flag is to limit the val loop

- sanity check
- train loop
- val loop

**Comment by williamFalcon** [comment]
But for this case, i guess if you wanted to run the full batch in sanity check, then -1 should be the flag that wins here.
Although... i honestly think the user should just remove the limit_val_batches flag...

Why ask to run the full val set for sanity check when you only want a portion of it in actual training?

**Comment by awaelchli** [comment]
Ok, I hope I understood you correctly. I updated the PR. Here are some examples of how it currently is:
```python
# runs exactly 5 sanity val steps
Trainer(num_sanity_val_steps=5)

# runs exactly 5 sanity val steps
Trainer(num_sanity_val_steps=5, limit_val_batches=3)

# runs 0 sanity val steps
Trainer(num_sanity_val_steps=5, limit_val_batches=0)

# runs 0 sanity val step
Trainer(num_sanity_val_steps=5, limit_val_batches=3, fast_dev_run=True)
```
Does this make sense?

**Comment by williamFalcon** [comment]
perfect!

**Comment by awaelchli** [comment]
yes exactly, I did not change that. I just did not conceptually count it as "sanity check" because in the code it follows a different path. (in fact, fast_dev_run sets num_sanity_val_steps=0 in Trainer init)

**Comment by awaelchli** [comment]
@williamFalcon kindly requesting your review :)

**Comment by williamFalcon** [comment]
@awaelchli this looks great. Can you rebase so we can merge?

File: pytorch_lightning/callbacks/progress.py [hunk 2]
```diff
@@ -302,7 +302,7 @@ def init_test_tqdm(self) -> tqdm:
     def on_sanity_check_start(self, trainer, pl_module):
         super().on_sanity_check_start(trainer, pl_module)
         self.val_progress_bar = self.init_sanity_tqdm()
-        self.val_progress_bar.total = trainer.num_sanity_val_steps * len(trainer.val_dataloaders)
+        self.val_progress_bar.total = convert_inf(trainer.num_sanity_val_steps * len(trainer.val_dataloaders))
```
Comment (williamFalcon) [inline pytorch_lightning/callbacks/progress.py:305]: what is this?

Comment (awaelchli) [reply pytorch_lightning/callbacks/progress.py:305]: tqdm does not understand float("inf"), so we have to convert it to None in the case num_sanity_val_steps=inf or dataloader has inf length.

**Comment by awaelchli** [comment]
@williamFalcon rebase done.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -883,6 +883,18 @@ def progress_bar_dict(self) -> dict:
         ref_model = self.model if not self.data_parallel else self.model.module
         return dict(**ref_model.get_progress_bar_dict(), **self.progress_bar_metrics)
 
+    @property
+    def disable_validation(self) -> bool:
+        """ Check if validation is disabled during training. """
+        disable_validation = not (self.is_overridden('validation_step') and self.limit_val_batches > 0) \
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:889]: when all are not, would it be easier to write `enable_validation` and negate it here? (vise verso to this state)

File: pytorch_lightning/callbacks/progress.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/callbacks/progress.py:305]: right but where is this imported?

Comment (williamFalcon) [reply pytorch_lightning/callbacks/progress.py:305]: oh i see... just kind of stuck at the bottom lol

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py]: done

**Comment by williamFalcon** [comment]
@awaelchli why does this drop coverage so much? maybe the logical part was wrong?

**Comment by awaelchli** [comment]
@williamFalcon it is because it fails to uplaod the coverage, http error.
